docs/supporter-data-flow.md

```diff
@@ -0,0 +1,105 @@
+
+Each ErpAction should validate against the following:
+
+```javascript
+{
+  "type": ErpPayloadTypes,
+  "data": {
+    // Upstream service specific data
+    specific_data_1: any,
+    specific_data_2: any,
+
+    // The following key is likely to be required
+    // for supporter-specific payloads
+    // such as donations, prize tickets, etc.
+    // See the `Supporter UUID` section for details
+    supporter_uuid: String,
+
+    // If your service doesn't have access
+    // to supporter specific information
+    // the payload will need to specify
+    // the `identity_hash` key so that `supporter-event-service`
+    // will be able to complete the payload
+    identity_hash: String
+  }
+}
+```
+
+See:
+- [ErpPayloadTypes](https://github.com/comicrelief/data-models/blob/master/src/schemas/erp/models/common.ts#ErpPayloadTypes)
+- [Erp Actions List](https://github.com/comicrelief/data-models/tree/master/src/schemas/erp/models/actions)
+
+## Supporter UUID
+
+Your ERP payloads may be related to a specific supporter. This is true for transactions, prize tickets, Shopify orders, etc. The `data.supporter_uuid` key will be used by ERP Next to link the entity with a Supporter entity. However `supporter_uuid` cannot be guaranteed to upstream services as it will be stored on `serverless-user-service`. This is why `supporter-event-service` can help your upstream service complete its payload.
+
+### Scenario 1: The upstream service provides a `user-service` specific payload
+
+As part of your payload to `supporter-event-service` you can specify a `user-service` specific payload. Should `supporter-event-service` receive a `user-service` payload, it will use it to create a user entity on `serverless-user-service`, if it doesn't already exist.
+
+The request to `serverless-user-service` will yield the user UUID, which will be used to populate all following payloads to `serverless-erp-service`, adding `data.supporter_uuid` if it isn't already defined.
+
+This means that the upstream service doesn't need to be concerned with `data.supporter_uuid` if it provides a `user-service` specific payload along the request.
+
+### Scenario 2: The upstream doesn't have access to supporter specific information
+
+Not all services may be storing supporter specific information and all they might have access to is an identity hash. This will be true if the service defines multiple endpoints that need to populate ERP Service with progressive data.
+
+`serverless-prize-platform` is an example of such a service. While creating a ticket entry, the lambda will have access to all the user data; but picking a winner among the tickets won't have access to any supporter information that wasn't stored within the service itself.
+
+`supporter-event-service` will need to handle this scenario as well. After all, it has direct access to `serverless-user-service`. Any payload requiring `data.supporter_uuid` can specify `data.identity_hash`.
```

Comment:
So what I've understood is: if you have a complete set of user details (full name, email, phone) you can put it in the `user-service` key and `supporter-event-service` will set `data.supporter_uuid` automatically. If you don't have all those details but do have an identity hash, you can put the identity hash in `data.identity_hash` to achieve the same thing.

Follow-up questions:

- Once `data.supporter_uuid` has been added, will `data.identity_hash` be deleted?

- What if I have more than one identity hash because the user gave me both an email address and phone number? Do I just pick one arbitrarily?

- It sounds like the precedent for Scenario 2 is that upstream services should not store raw user details, but should store only the identity hashes (or only one of them?) returned by the initial request to `supporter-event-service`. If that's correct, could we state it somewhere?

Reply:
Very good points, thanks @seb-cr 
98a0cae08c10f7fda6c2b8196a82727822177478

Reply:
That's great, thanks!